File: src/Sample/Flutter.Net/StackDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Flutter.Rendering;
using Flutter.Widgets;

// Dart parity source (reference): dart_sample/lib/stack_demo_page.dart (exact sample parity)

namespace Flutter.Net;

public sealed class StackDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new StackDemoPageState();
    }
}

internal sealed class StackDemoPageState : State
{
    private double _left = 8;
    private double _top = 8;
    private bool _pinBottomRight;

    public override Widget Build(BuildContext context)
    {
        Widget badge = new Positioned(
            left: _pinBottomRight ? null : _left,
            top: _pinBottomRight ? null : _top,
            right: _pinBottomRight ? 8 : null,
            bottom: _pinBottomRight ? 8 : null,
            child: new Container(
                width: 56,
                height: 30,
                color: Color.Parse("#FFD1495B"),
                child: new Center(
                    child: new Text("badge", fontSize: 11, color: Colors.White))));

        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("Stack + Positioned", fontSize: 20, color: Colors.Black),
                new Text(
                    "Mix non-positioned and positioned children; toggle pinned mode and nudge offsets.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("Left", () => Move(-8, 0), width: 72, colorHex: "#FFDCE3ED"),
                        BuildButton("Right", () => Move(8, 0), width: 72, colorHex: "#FFDCE3ED"),
                        BuildButton("Up", () => Move(0, -8), width: 72, colorHex: "#FFDCE3ED"),
                        BuildButton("Down", () => Move(0, 8), width: 72, colorHex: "#FFDCE3ED"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton(_pinBottomRight ? "Pin: BR" : "Pin: custom", TogglePin, width: 120, colorHex: "#FFE9F5EC"),
                        BuildButton("Reset", Reset, width: 88, colorHex: "#FFF3E8D8"),
                    ]),
                new Text(
                    $"left={_left:0}, top={_top:0}, mode={(_pinBottomRight ? "bottomRight" : "custom")}",
                    fontSize: 12,
                    color: Colors.DarkSlateGray),
                new Container(
                    width: 220,
                    height: 140,
                    color: Color.Parse("#FFE7EDF6"),
                    padding: new Thickness(8),
                    child: new Container(
                        color: Colors.White,
                        child: new Stack(
                            alignment: Alignment.Center,
                            children:
                            [
                                new Container(
                                    width: 140,
                                    height: 80,
                                    color: Color.Parse("#FFCCE3FF"),
                                    child: new Center(
                                        child: new Text("base", fontSize: 14, color: Color.Parse("#FF1D3557")))),
                                badge,
                            ]))),
            ]);
    }

    private Widget BuildButton(string label, Action onTap, double width, string colorHex)
    {
        return new SizedBox(
            width: width,
            child: new CounterTapButton(
                label: label,
                onTap: onTap,
                background: Color.Parse(colorHex),
                foreground: Colors.Black,
                fontSize: 12,
                padding: new Thickness(10, 8)));
    }

    private void Move(double dx, double dy)
    {
        if (_pinBottomRight)
        {
            return;
        }

        SetState(() =>
        {
            _left = Math.Clamp(_left + dx, 0, 150);
            _top = Math.Clamp(_top + dy, 0, 90);
        });
    }

    private void TogglePin()
    {
        SetState(() => _pinBottomRight = !_pinBottomRight);
    }

    private void Reset()
    {
        SetState(() =>
        {
            _left = 8;
            _top = 8;
            _pinBottomRight = false;
        });
    }
}

File: dart_sample/lib/stack_demo_page.dart
import 'package:flutter/material.dart';

import 'counter_widgets.dart';

class StackDemoPage extends StatefulWidget {
  const StackDemoPage({super.key});

  @override
  State<StackDemoPage> createState() => _StackDemoPageState();
}

class _StackDemoPageState extends State<StackDemoPage> {
  double _left = 8;
  double _top = 8;
  bool _pinBottomRight = false;

  @override
  Widget build(BuildContext context) {
    final Widget badge = Positioned(
      left: _pinBottomRight ? null : _left,
      top: _pinBottomRight ? null : _top,
      right: _pinBottomRight ? 8 : null,
      bottom: _pinBottomRight ? 8 : null,
      child: Container(
        width: 56,
        height: 30,
        color: const Color(0xFFD1495B),
        child: const Center(
          child: Text(
            'badge',
            style: TextStyle(fontSize: 11, color: Colors.white),
          ),
        ),
      ),
    );

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'Stack + Positioned',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Mix non-positioned and positioned children; toggle pinned mode and nudge offsets.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'Left',
              onTap: () => _move(-8, 0),
              width: 72,
              background: const Color(0xFFDCE3ED),
            ),
            _buildButton(
              label: 'Right',
              onTap: () => _move(8, 0),
              width: 72,
              background: const Color(0xFFDCE3ED),
            ),
            _buildButton(
              label: 'Up',
              onTap: () => _move(0, -8),
              width: 72,
              background: const Color(0xFFDCE3ED),
            ),
            _buildButton(
              label: 'Down',
              onTap: () => _move(0, 8),
              width: 72,
              background: const Color(0xFFDCE3ED),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: _pinBottomRight ? 'Pin: BR' : 'Pin: custom',
              onTap: _togglePin,
              width: 120,
              background: const Color(0xFFE9F5EC),
            ),
            _buildButton(
              label: 'Reset',
              onTap: _reset,
              width: 88,
              background: const Color(0xFFF3E8D8),
            ),
          ],
        ),
        Text(
          'left=${_left.toStringAsFixed(0)}, top=${_top.toStringAsFixed(0)}, mode=${_pinBottomRight ? 'bottomRight' : 'custom'}',
          style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
        ),
        Container(
          width: 220,
          height: 140,
          color: const Color(0xFFE7EDF6),
          padding: const EdgeInsets.all(8),
          child: Container(
            color: Colors.white,
            child: Stack(
              alignment: Alignment.center,
              children: <Widget>[
                Container(
                  width: 140,
                  height: 80,
                  color: const Color(0xFFCCE3FF),
                  child: const Center(
                    child: Text(
                      'base',
                      style: TextStyle(fontSize: 14, color: Color(0xFF1D3557)),
                    ),
                  ),
                ),
                badge,
              ],
            ),
          ),
        ),
      ],
    );
  }

  Widget _buildButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: CounterTapButton(
        label: label,
        onTap: onTap,
        background: background,
        foreground: Colors.black,
        fontSize: 12,
        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
      ),
    );
  }

  void _move(double dx, double dy) {
    if (_pinBottomRight) {
      return;
    }

    setState(() {
      _left = (_left + dx).clamp(0.0, 150.0);
      _top = (_top + dy).clamp(0.0, 90.0);
    });
  }

  void _togglePin() {
    setState(() {
      _pinBottomRight = !_pinBottomRight;
    });
  }

  void _reset() {
    setState(() {
      _left = 8;
      _top = 8;
      _pinBottomRight = false;
    });
  }
}
